libs/langchain_v1/langchain/agents/middleware/pii.py

```diff
@@ -0,0 +1,696 @@
+            masked = (
+                f"**{separator}**{separator}**{separator}**{separator}**{separator}{value[-2:]}"
+            )
+
+        elif pii_type == "url":
+            # Mask everything: [MASKED_URL]
+            masked = "[MASKED_URL]"
+
+        else:
+            # Default: show last 4 chars
+            masked = f"****{value[-4:]}" if len(value) > 4 else "****"
+
+        result = result[: match["start"]] + masked + result[match["end"] :]
+
+    return result
+
+
+def _apply_hash_strategy(content: str, matches: list[PIIMatch]) -> str:
+    """Replace PII with deterministic hash including type information.
+
+    Args:
+        content: Original content.
+        matches: List of PII matches to hash.
+
+    Returns:
+        Content with PII replaced by hashes in format <type_hash:digest>.
+    """
+    if not matches:
+        return content
+
+    # Sort matches by start position in reverse
+    sorted_matches = sorted(matches, key=lambda m: m["start"], reverse=True)
+
+    result = content
+    for match in sorted_matches:
+        value = match["value"]
+        pii_type = match["type"]
+        # Create deterministic hash
+        hash_digest = hashlib.sha256(value.encode()).hexdigest()[:8]
+        replacement = f"<{pii_type}_hash:{hash_digest}>"
+        result = result[: match["start"]] + replacement + result[match["end"] :]
+
+    return result
+
+
+# ============================================================================
+# PIIMiddleware
+# ============================================================================
+
+
+class PIIMiddleware(AgentMiddleware):
+    """Detect and handle Personally Identifiable Information (PII) in agent conversations.
+
+    This middleware detects common PII types and applies configurable strategies
+    to handle them. It can detect emails, credit cards, IP addresses,
+    MAC addresses, and URLs in both user input and agent output.
+
+    Built-in PII types:
+        - ``email``: Email addresses
+        - ``credit_card``: Credit card numbers (validated with Luhn algorithm)
+        - ``ip``: IP addresses (validated with stdlib)
+        - ``mac_address``: MAC addresses
+        - ``url``: URLs (both http/https and bare URLs)
+
+    Strategies:
+        - ``block``: Raise an exception when PII is detected
+        - ``redact``: Replace PII with ``[REDACTED_TYPE]`` placeholders
+        - ``mask``: Partially mask PII (e.g., ``****-****-****-1234`` for credit card)
+        - ``hash``: Replace PII with deterministic hash (e.g., ``<email_hash:a1b2c3d4>``)
+
+    Strategy Selection Guide:
+
+        ========  ===================  =======================================
+        Strategy  Preserves Identity?  Best For
+        ========  ===================  =======================================
+        `block`   N/A                  Avoid PII completely
+        `redact`  No                   General compliance, log sanitization
```

Comment:
@nfcampos I'll need to document that this is not enough for PII compliance likely as users will also need to make sure that they can sanitize anything associated with their own logging set up (kidna obvious, but maybe not to everyone) + langgraph checkpointer?